test/unit/communication/test_transform_mpi.cpp

```diff
@@ -0,0 +1,131 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2022, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#include "dlaf/sender/transform_mpi.h"
+
+#include <atomic>
+#include <chrono>
+#include <thread>
+
+#include <gtest/gtest.h>
+
+#include "dlaf/common/pipeline.h"
+#include "dlaf/communication/communicator.h"
+#include "dlaf/sender/when_all_lift.h"
+
+using namespace dlaf;
+using namespace dlaf::comm;
+
+using TransformMPITest = ::testing::Test;
+
+// wait for guard to become true with a timeout of 200ms
+auto try_waiting_guard = [](std::atomic_bool& guard) {
+  using namespace std::chrono_literals;
+  const auto wait_guard = 20ms;
+
+  for (int i = 0; i < 100 && !guard; ++i)
+    std::this_thread::sleep_for(wait_guard);
+};
+
+TEST_F(TransformMPITest, PromiseGuardManagement) {
+  static_assert(NUM_MPI_RANKS >= 2, "This test requires 2 ranks");
+
+  // Note:
+  // Rank 0 is where the test actually is done, while Rank 1 is just used by Rank 0 to control
+  // its own progress.
+  //
+  // Rank 0 posts an IRecv that will be satisfied by Rank 1 just after it receives from Rank 0
+  // the trigger. So, actually, Rank 0 can control when to unlock its IRecv by deciding when to
+  // send the trigger to Rank 1. In this way, Rank 0 flow can be "paused" for verifiying various
+  // phases of the async communication mechanism for IRecv.
+
+  Communicator world(MPI_COMM_WORLD);
+  const IndexT_MPI rank = world.rank();
+
+  if (rank == 0) {
+    using dlaf::comm::internal::transformMPI;
+    using dlaf::internal::whenAllLift;
+
+    namespace ex = pika::execution::experimental;
+
+    common::Pipeline<Communicator> chain(world);
```

Comment:
Communicator `world` gets passed into Pipeline without being moved, and it gets used later for the signal message.

Copying a Communicator means copying the handle, so we end up using the same handle from different threads (in edge cases concurrently). It should be ok, since MPI calls should be thread-safe, moreover it is the `MPI_COMM_WORLD` communicator which is managed by MPI directly (so there are no additional caveats about lifetime management).

In case it is not, we may opt for cloning the Communicator on both ranks and using one for the `IRecv` communication and one for the signal part.

Let me know what you would prefer.

Reply:
No need to duplicate in this case as only two messages that have a unique match are exchanged.